# Content list

1. Import libraries
2. Import data
3. Deriving new variables
4. Task 4.7
5. Export data

# 01. Import libraries

In [1]:
#Import libraries

import pandas as pd
import numpy as np
import os

# 02. Import data

In [2]:
# Create path to main folder

path = r'C:\Users\adrig\OneDrive\Documentos\CARRIERFOUNDRY\DATA IMMERSION\ACHIEVEMENT 4\Ex_4.2_Python\03-2025 Instacart Basket Analysis'

In [3]:
# Import orders_products_merged dataset

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [4]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_last_order  new_customer  product_id  add_to_cart_order  \
0                    NaN          True         196                  1   
1                    NaN          True       14084                  2   
2                    NaN          True       12427                  3   
3                    NaN          True       26088                  4   
4                    NaN          True       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices  
0              7     9.0  
1             16    12.5  
2             19     4.4  
3             19     4.7  
4             17     1.0

In [5]:
# Create a subset with the first 1,000,000 rows

df = ords_prods_merge[:1000000]

In [6]:
df.shape

(1000000, 14)

## 03. Deriving new variables

## 03.1 If statements with user-defined functions

In [7]:
# Define the function 'def' followed by the name 'price_label'  
# The argument (row) #looks at each row within the dataframe and : separates the head from the body

def price_label(row):  #

    if row['prices'] <= 5:
        return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
          return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High-range product'
    else: return 'Not enough data'

In [8]:
# Apply the function

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\adrig\AppData\Local\Temp\ipykernel_26244\93659844.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [9]:
# Count values for the new column

df['price_range'].value_counts(dropna = False)

price_range
Mid-range product     673340
Low-range product     314392
High-range product     12268
Name: count, dtype: int64

In [10]:
df['prices'].max()

25.0

In [11]:
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_last_order  new_customer  product_id  add_to_cart_order  \
0                    NaN          True         196                  1   
1                    NaN          True       14084                  2   
2                    NaN          True       12427                  3   
3                    NaN          True       26088                  4   
4                    NaN          True       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices        price_range  
0              7     9.0  Mid-range product  
1             16    12.5  Mid-range product  
2             19     4.4  Low-range product  
3             19     4.7  Low-range product  
4             17     1.0  Low-range product

## 03.2 If statements with the loc() function

In [12]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High range product'

C:\Users\adrig\AppData\Local\Temp\ipykernel_26244\2555204669.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High range product'


In [13]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid range product'

In [14]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low range product'

In [15]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid range product     673340
Low range product     314392
High range product     12268
Name: count, dtype: int64

In [16]:
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_last_order  new_customer  product_id  add_to_cart_order  \
0                    NaN          True         196                  1   
1                    NaN          True       14084                  2   
2                    NaN          True       12427                  3   
3                    NaN          True       26088                  4   
4                    NaN          True       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices        price_range    price_range_loc  
0              7     9.0  Mid-range product  Mid range product  
1             16    12.5  Mid-range product  Mid range product  
2             19     4.4  Low-range product  Low range product  
3             19     4.7  Low-range product  Low range product  
4             17     1.0  Low-range product  Low range product

Creating the new column in the entire dataframe

In [17]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [18]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [19]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [20]:
ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21865987
Low-range product     10126321
High-range product      412551
Name: count, dtype: int64

## 03.3 If statements with for loops

In [21]:
# loop cycles for every value x within the range of 30 and 45
for x in range(30,45):
    print('My age is %d' % (x))

My age is 30
My age is 31
My age is 32
My age is 33
My age is 34
My age is 35
My age is 36
My age is 37
My age is 38
My age is 39
My age is 40
My age is 41
My age is 42
My age is 43
My age is 44


In [22]:
# Find out on which days most orders take place

ords_prods_merge['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [23]:
# Create the statement for a new column 'Busiest day'

result = [] # This is to create an empty list or shell for the results

for value in ords_prods_merge['orders_day_of_week']:
    if value == 0:
        result.append('Busiest day')
    elif value == 4:
        result.append('Least busy')
    else:
        result.append('Regularly busy')

In [24]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy

In [25]:
# Create a new column to combine results with dataframe

ords_prods_merge['busiest_day'] = result

In [26]:
ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

In [27]:
ords_prods_merge['busiest_day'].count()

32404859

In [28]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_last_order  new_customer  product_id  add_to_cart_order  \
0                    NaN          True         196                  1   
1                    NaN          True       14084                  2   
2                    NaN          True       12427                  3   
3                    NaN          True       26088                  4   
4                    NaN          True       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices    price_range_loc     busiest_day  
0              7     9.0  Mid-range product  Regularly busy  
1             16    12.5  Mid-range product  Regularly busy  
2             19     4.4  Low-range product  Regularly busy  
3             19     4.7  Low-range product  Regularly busy  
4             17     1.0  Low-range product  Regularly busy

# 04. Task 4.7

## 04.2 New column busiest days (plural)

In [29]:
# Create the statement for the new column

result_days = [] # This is to create an empty list or shell for the results

for value in ords_prods_merge['orders_day_of_week']:
    if value == 0 or value == 1:
        result_days.append('Busiest days')
    elif value == 3 or value == 4:
        result_days.append('Slowest days')
    else:
        result_days.append('Regular days')

In [30]:
result_days

['Regular days',
 'Regular days',
 'Regular days',
 'Regular days',
 'Regular days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regular days',
 'Regular days',
 'Regular days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days

In [31]:
# Create column for result_days

ords_prods_merge['busiest_days'] = result_days

In [32]:
# Frequency of new column

ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regular days    12916111
Busiest days    11864412
Slowest days     7624336
Name: count, dtype: int64

In [33]:
ords_prods_merge['busiest_days'].count()

32404859

In [34]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_last_order  new_customer  product_id  add_to_cart_order  \
0                    NaN          True         196                  1   
1                    NaN          True       14084                  2   
2                    NaN          True       12427                  3   
3                    NaN          True       26088                  4   
4                    NaN          True       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices    price_range_loc     busiest_day  busiest_days  
0              7     9.0  Mid-range product  Regularly busy  Regular days  
1             16    12.5  Mid-range product  Regularly busy  Regular days  
2             19     4.4  Low-range product  Regularly busy  Regular days  
3             19     4.7  Low-range product  Regularly busy  Regular days  
4             17     1.0  Low-range product  Regularly busy  Regular days

## 04.3 Check the values for the new column

In [35]:
ords_prods_merge.count()

order_id                 32404859
user_id                  32404859
order_number             32404859
orders_day_of_week       32404859
order_hour_of_day        32404859
days_since_last_order    30328763
new_customer             32404859
product_id               32404859
add_to_cart_order        32404859
reordered                32404859
product_name             32404859
aisle_id                 32404859
department_id            32404859
prices                   32404859
price_range_loc          32404859
busiest_day              32404859
busiest_days             32404859
dtype: int64

The column was succesfully created, the total counts of the column busiest_days match with the total counts of the previously created column busiest_day, and the total counts from the dataframe (32404859).  

## 04.4 Busiest hours of the day (periods of time)

In [36]:
# Find out which hours of the day have the most orders 

ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

Looking at the values, I decided to divide the periods of time according to the amount of orders: Most_orders will be the hours with >1,000,000 orders (10,11,14,15,13,12,16,9,17,8,18,19), Average_orders for the hours with >100,000 and <1,000,000 (20,7,21,22,23,6,0,1) and Fewest_orders will be the hours with >100,000 orders (5,2,4,3). 

In [37]:
# Create the statement for the new column

result_hours = []

# Define values group

group_most = [10,11,14,15,13,12,16,9,17,8,18,19]
group_average = [20,7,21,22,23,6,0,1]

# Define the loop function 

for value in ords_prods_merge['order_hour_of_day']:
    if value in group_most:
        result_hours.append('Most_orders')
    elif value in group_average:
        result_hours.append('Average_orders')
    else:
        result_hours.append('Fewest_orders')

In [38]:
result_hours

['Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Average_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orders',
 'Most_orde

In [39]:
# Create column for result_days

ords_prods_merge['busiest_period_of_day'] = result_hours

In [40]:
# Frequency of new column

ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most_orders       27818650
Average_orders     4324350
Fewest_orders       261859
Name: count, dtype: int64

In [41]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_last_order  new_customer  product_id  add_to_cart_order  \
0                    NaN          True         196                  1   
1                    NaN          True       14084                  2   
2                    NaN          True       12427                  3   
3                    NaN          True       26088                  4   
4                    NaN          True       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices    price_range_loc     busiest_day  busiest_days  \
0              7     9.0  Mid-range product  Regularly busy  Regular days   
1             16    12.5  Mid-range product  Regularly busy  Regular days   
2             19     4.4  Low-range product  Regularly busy  Regular days   
3             19     4.7  Low-range product  Regularly busy  Regular days   
4             17     1.0  Low-range product  Regularly busy  Regular days   

  busiest_period_of_day  
0           Most_orders  
1           Most_orders  
2           Most_orders  
3           Most_orders  
4           Most_orders

# 05. Export data

In [42]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_new_var.pkl'))